Add pickling of Atom, Residue, Segment and groups.

pgbarletta opened the pull request:
From #3947

Changes made in this Pull Request:

- Add reduce() method to both ResidueGroup and SegmentGroup 
- Change _unpickle() so it can create any kind of group from its indices. 
- Add one test for each group in the parallelism section.


PR Checklist
------------
 - [X] Tests?
 - [x] Docs?
 - [X] CHANGELOG updated?
 - [X] Issue raised/referenced?

PD: sorry for the new PR. I synced up my fork from develop and that somehow closed the previous PR.

richardjgowers commented on package/MDAnalysis/core/groups.py:
```diff
@@ -115,8 +115,15 @@
 
-def _unpickle(u, ix):
-    return u.atoms[ix]
+def _unpickle(u, ix, cls=None):
+    if cls:
+        group = cls(ix, u)
+    else:
+        group = u.atoms[ix]
+        warnings.warn("You may be trying to read an AtomGroup that was pickled "
+        "with a pre 2.4 MDAnalysis version. If so, rewrite the pickle file to "
+        "silence this warning. More info at issue #3947 and PR #3953.", DeprecationWarning)
+    return group
```
so a different solution to this would be to keep AtomGroup pickling going through this function (and not break anything) and create a new function (`_unpickle2`) which has the new signature.  This can then get tidied in 3.0 (which we *can* break compatibility with).

then you won't have the untested branch in the code

pgbarletta replied:
Done.

richardjgowers commented on testsuite/MDAnalysisTests/core/test_segmentgroup.py:
```diff
@@ -109,3 +111,16 @@ def test_set_segids_many():
+    """Test SegmentGroup pickling support."""
+    @pytest.fixture()
+    def universe(self):
+        return mda.Universe(TPR, XTC)
```
minor quibble, but if you use `Universe.empty()` it'll make the tests a little faster.  it's not a big deal, but over 1,000 tests it adds up slowly

pgbarletta replied:
Sorry, don't get it. Say I do:
```
seg_group = mda.Universe.empty(10).segments
seg = pickle.loads(pickle.dumps(seg_group))
```

how do I compare them? And would that be a useful test?

richardjgowers replied:
you could check that the new segments have the same universe as the old one:
```python
u = mda.Universe.empty()

segs = u.segments

new = pickle.loads(...)

assert new.universe is u
```

I think maybe you're right, a better test would have dummy data inside the Universe...
```python
u = mda.Universe.empty()
u.add_TopologyAttr('segids', values=['A', 'B', etc]

... some pickling

assert new.segids == ['A', 'B', etc]
```

The idea is you can create a Universe without using files, and it'll make the tests faster if we don't (and also makes the tests not reliant on TPRParser working)

pgbarletta replied:
thanks. I went with the 2nd choice:
```
def test_segmentgroup_pickle():
    u = mda.Universe.empty(10)
    u.add_Segment(segid="X")
    u.add_Segment(segid="Y")
    u.add_Segment(segid="Z")
    segids = ["A", "X", "Y", "Z"]
    u.add_TopologyAttr("segids", values=["A", "X", "Y", "Z"])
    seg_group = mda.SegmentGroup((1, 3), u)
    seg = pickle.loads(pickle.dumps(seg_group))
    assert_equal(seg.universe.segments.segids, segids)
```

richardjgowers reviewed (changes requested):
can we use empty Universes?

hmacdope commented on package/MDAnalysis/core/groups.py:
```diff
@@ -119,6 +119,10 @@ def _unpickle(u, ix):
     return u.atoms[ix]
 
 
+def _unpickle2(u, ix, cls):
```
Just one little nitpick, would you be able to document in a quick comment why the two are necessary for someone following along later?

hmacdope commented on package/CHANGELOG:
```diff
@@ -39,6 +39,8 @@ Fixes
   * fixes guessed_attributes and read_attributes methods of the Topology class, as 
     those methods were giving unexpected results for some topology attributes
     (e.g. bonds, angles) (PR #3779).
+  * Add pickling support for Atom, Residue, Segment, ResidueGroup 
```
Entries newest first and this needs to be put under 2.5.0. Merging with upstream/develop also required. :)

hmacdope reviewed (changes requested):
Two little things and then I think good to go.

hmacdope commented on package/AUTHORS:
```diff
@@ -199,6 +199,7 @@ Chronological list of authors
   - Haleema Khan
   - Jennifer A Clark
   - Jake Fennick
+  - Patricio Barletta
```
Nitpick but you should be last

hmacdope reviewed (changes requested):
Two tiny nitpicks sorry.

hmacdope commented on package/CHANGELOG:
```diff
@@ -21,6 +21,9 @@ Fixes
 
 Enhancements
 
+* Add pickling support for Atom, Residue, Segment, ResidueGroup 
```
Add your self to 2.5.0 authors to the right of IAlibay. :)

pgbarletta commented:
I'll be out for a ~week. I'll rebase once the numpy deprecations issues are fixed.

hmacdope reviewed (approved):
LGTM thanks @pgbarletta

IAlibay commented on package/CHANGELOG:
```diff
@@ -21,12 +21,15 @@ Fixes
 
 
-12/21/22 IAlibay
+12/21/22 IAlibay, pgbarletta
```
```suggestion
12/21/22 IAlibay
```
Not a 2.4.2 change? (edit: I think the intention here was to add this to line 16 instead?)

pgbarletta replied:
Fixed.

pgbarletta committed:
Add pickling of Atom, Residue, Segment and groups.

Add reduce() method to both ResidueGroup and SegmentGroup
Change _unpickle() so it can create any kind of group from its indices.
Add one test for each group in the parallelism section.

IAlibay commented on package/MDAnalysis/core/groups.py:
```diff
@@ -119,6 +119,11 @@ def _unpickle(u, ix):
     return u.atoms[ix]
 
 
+# TODO 3.0: deprecate _unpickle in favor of _unpickle2.
```
Just a comment here, unless it changes user-facing behaviour there is no need to deprecate a private method like this. We can just throw it out right now and none would be the wiser.

Please raise a relevant issue regarding this though.

pgbarletta replied:
Done.

IAlibay reviewed (approved):
I'll approve although please do raise the issue for the follow-up 3.0 work, otherwise it'll get lost.
